### ISSUE TITLE
Cases within match statement do not exhaustively handle all values   Unhandled type: "." when using unions

### ISSUE THREAD

**Issue by patrick91** [microsoft/pylance-release#4692]
<h3> Environment data </h3>

-   Language Server version: 2023.8.10
-   OS and version: darwin arm64
-   Python version (and distribution if applicable, e.g. Anaconda): 
-   python.analysis.indexing: true
-   python.analysis.typeCheckingMode: strict

<h3> Code Snippet </h3>

```python
import json
from typing import Literal, TypedDict, Any, cast


class CodegenMessage(TypedDict):
    type: Literal["message"]
    message: str


class CodegenResult(TypedDict):
    type: Literal["result"]
    sandbox_url: str


CodegenChunk = CodegenMessage | CodegenResult


def something(data: Any) -> None:
    result = cast(CodegenChunk, json.loads(data["message"]))

    match result:
        case {"type": "message", "message": text}:
            print("message", text)
        case {"type": "result", "sandbox_url": sandbox_url}:
            print("result", sandbox_url)

```

<h3> Repro Steps </h3>

Run VS code on that file with pylance
1. XXX

<h3> Expected behavior </h3>

No Error

<h3> Actual behavior </h3>

```
Cases within match statement do not exhaustively handle all values
  Unhandled type: "CodegenChunk"
  If exhaustive handling is not intended, add "case _: pass"Pylance[reportMatchNotExhaustive](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportMatchNotExhaustive)
```

<img width="609" alt="CleanShot 2023-08-07 at 12 40 57@2x" src="https://github.com/microsoft/pylance-release/assets/667029/84fbcf17-b9e1-459e-82f3-6d615359f7d6">

### Notes

This issue not present on pyright

**Comment by erictraut** [comment]
It looks like you have the `reportMatchNotExhaustive` check enabled. This check isn't enabled by default. It's considered a "strict" type checking option, and it's a bit opinionated in how it works.

Exhaustion detection is based on type narrowing. You're using a pattern here that pyright (the type checker that underlies pylance) does not narrow. Mypy (another Python static type checker) also doesn't support this pattern. What you're doing here is pretty atypical. The more typical way of handling this problem is to match on the discriminating key only. Both pyright and mypy support this.

If you change your code as follows, it will work as expected.
```python
def something(data: Any) -> None:
    result = cast(CodegenChunk, json.loads(data["message"]))

    match result:
        case {"type": "message"}:
            print("message", result["message"])
        case {"type": "result"}:
            print("result", result["sandbox_url"])
```

I don't know how this function is intended to be used, but if the JSON data is coming from some external source (i.e. its contents haven't been generated from within your program or already validated by some other code), then it's not safe to assume that its contents will match one of these two patterns. In that case, I recommend that you stick with the original code but include a fallback pattern:

```python
    match result:
        case {"type": "message", "message": text}:
            print("message", text)
        case {"type": "result", "sandbox_url": sandbox_url}:
            print("result", sandbox_url)
        case _:
            print("None of the above")
```

In summary, I don't consider this a bug in pyright. It could be a feature request, but we're unlikely to add support for this particular type narrowing pattern unless/until we get signal from enough other pyright users that it's a commonly-used pattern.

**Comment by rchiodo** [comment]
I think we'd rather close it. If it's not actually a bug we don't keep issues open.
